### PR TITLE
Mention which users have not signed the CLA

### PR DESCRIPTION
**Pull request by potomak**
Update `check` and `update` to include in the message a list of users
that either didn't sign the CLA or haven't been found.

The implementation follows [@brettcannon's suggestion](https://github.com/python/the-knights-who-say-ni/issues/128#issuecomment-337313178)
to return a `Mapping[Status, AbstractSet[str]]` from `check` that is a
map of "problematic" statuses and their associated list of usernames.

See #128

### CODE REVIEW

File: ni/test/test_bpo.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def get(self, url):
         return self
 
     async def __aenter__(self):
-        self.session = aiohttp.ClientSession(loop=self.loop)
+        self.session = aiohttp.ClientSession(loop=self.loop, trust_env=True)
```
Comment (potomak) [inline ni/test/test_bpo.py]: Note: I had to add this option in order to be able to use env variables for setting a proxy, see https://docs.aiohttp.org/en/stable/client_advanced.html?highlight=proxy#proxy-support.

I can remove it if you think we shouldn't trust env variables (even if we're just testing).

**Comment by potomak** [comment]
Note: after this update `Status` doesn't need to include the `signed` value anymore because it's used only as a list of "problematic" statuses :)

I think we could

1. rename `Status` to `Problem`
2. remove from the enum the `signed` value

What do you think about this update?

If you agree with this update, would it make sense to include it in this PR?

File: ni/abc.py [hunk 2]
```diff
@@ -101,10 +101,9 @@ class CLAHost(abc.ABC):
...
-        # mapping of {username: Status} was returned, the vast majority of
-        # cases will be for a single user and thus not worth the added
-        # complexity to need to worry about it.
-        return Status.username_not_found  # pragma: no cover
+                    usernames: AbstractSet[str]) -> Mapping[Status, AbstractSet[str]]:
+        """Check if all of the specified usernames have signed the CLA.
+
+        Return a Mapping of problems and the associated list of usernames.
+        """
+        return {Status.username_not_found: {'username'}}  # pragma: no cover
```
Comment (brettcannon) [inline ni/abc.py]: This doesn't make sense anymore as a default value, so it would probably be better to raise `NotImplementedError`.

File: ni/bpo.py [hunk 3]
```diff
@@ -35,9 +36,14 @@ def __init__(self, server: ni_abc.ServerHost) -> None:
...
-            return ni_abc.Status.username_not_found
-        else:
-            return ni_abc.Status.not_signed
+        failures = {
+            None: ni_abc.Status.username_not_found,
+            False: ni_abc.Status.not_signed,
+        }
+        problems: Mapping[ni_abc.Status, Set[str]] = defaultdict(set)
+        for username, result in results.items():
+            if result:
```
Comment (brettcannon) [inline ni/bpo.py]:
```suggestion
            if result not in failures:
```

File: ni/bpo.py [hunk 4]
```diff
@@ -35,9 +36,14 @@ def __init__(self, server: ni_abc.ServerHost) -> None:
...
+                continue
+            problems[failures[result]].add(username)
```
Comment (brettcannon) [inline ni/bpo.py]: Another way to do this which alleviates the need for `defaultdict` is:
```suggestion
            problems.setdefault(failures[result], set()).add(username)
```
This way if anyone chooses to rely on `KeyError` to detect any potential failures it will continue to work as expected versus having to check if the returned set if empty.

File: ni/github.py [hunk 5]
```diff
@@ -174,15 +181,15 @@ def __init__(self, server: ni_abc.ServerHost, client: aiohttp.ClientSession,
...
 
-    async def set_label(self, status: ni_abc.Status) -> str:
+    async def set_label(self, problems: Mapping[ni_abc.Status, AbstractSet[str]]) -> str:
         """Set the label on the pull request based on the status of the CLA."""
         labels_url = await self.labels_url()
-        if status == ni_abc.Status.signed:
-            await self._gh.post(labels_url, data=[CLA_OK])
-            return CLA_OK
-        else:
+        if problems:
```
Comment (brettcannon) [inline ni/github.py]: This is a good example of why `defaultdict` would be troublesome because if I had simply checked for `Status.not_signed` then this would have been considered true even if the set for `Status.not_signed` was empty.

File: ni/test/test_bpo.py (hunk 1 above)
Comment (brettcannon) [reply ni/test/test_bpo.py]: Yeah, please remove it just to be safe.

File: ni/abc.py [hunk 6]
```diff
@@ -101,10 +101,9 @@ class CLAHost(abc.ABC):
 
     @abc.abstractmethod
     async def check(self, client: aiohttp.ClientSession,
```
Comment (brettcannon) [inline ni/abc.py]: This should get renamed since `not check()` doesn't clearly convey that  there's no CLA issues. Otherwise successful checks should be included in the dict.

File: ni/test/test_github.py [hunk 7]
```diff
@@ -249,42 +249,69 @@ def test_comment(self):
...
         session = util.FakeSession(responses)
         contrib = github.Host(util.FakeServerHost(),
                               session,
                               github.PullRequestEvent.opened,
                               self.opened_example)
-        message = self.run_awaitable(contrib.comment(ni_abc.Status.username_not_found))
+        message = self.run_awaitable(contrib.comment({ni_abc.Status.username_not_found: {'username'}}))
+        self.assertEqual(expected['body'], message)
+
+        # Test the edge case when multiple users didn't sign the CLA and are not found
```
Comment (brettcannon) [inline ni/test/test_github.py]:
```suggestion
        # Test for when multiple users didn't sign the CLA and are yet others are not found.
```

File: ni/bpo.py (hunk 3 above)
Comment (potomak) [reply ni/bpo.py]: Yes, much better.

File: ni/bpo.py (hunk 4 above)
Comment (potomak) [reply ni/bpo.py]: Ah, I must admit I didn't remember about `setdefault`, thanks for the feedback.

File: ni/abc.py (hunk 6 above)
Comment (potomak) [reply ni/abc.py]: What name do you suggest?

1. `validate`
1. `find_problems`

Comment (brettcannon) [reply ni/abc.py]: `problems`?

**Comment by potomak** [comment]
@brettcannon thanks for reviewing this PR.

I addressed all the comments. Let me know if there's anything else to do.

**Comment by brettcannon** [comment]
@potomak I just want to let you know that I'm not ignoring this, I'm just really busy at the moment. Unfortunately I can't make any promises as to when I will get around to this PR, but it is in my review queue so I won't lose track of it.

**Comment by potomak** [comment]
@brettcannon thanks for the update.

> Unfortunately I can't make any promises as to when I will get around to this PR, but it is in my review queue

It would be interesting to see the size of your review queue + review time probability distribution conditional on some other parameters such as project and significant lines updated, in order to have an estimated ETA on the review.

**Comment by brettcannon** [comment]
Thanks for the PR! (And sorry for the delay for this final review.)
